third_party/move/extensions/async/move-async-vm/src/async_vm.rs

```diff
@@ -77,12 +77,12 @@ impl AsyncVM {
     }
 
     /// Creates a new session.
-    pub fn new_session<'r, 'l, S: MoveResolver>(
+    pub fn new_session<'r, 'l>(
         &'l self,
         for_actor: AccountAddress,
         virtual_time: u128,
-        move_resolver: &'r mut S,
-    ) -> AsyncSession<'r, 'l, S> {
+        move_resolver: &'r mut dyn MoveResolver,
```

Comment:
not sure if I understand why this can't be impl? how does it reduce the dependency?

Reply:
The moment you go through a dyn using dynamic dispatch, you lose syntactic dependence on the exact type. When using impl the code still depends on the yet unknown implementation of S: MoveResolver. This means that the compilation happens when the function is instantiated with a concrete type. The actual compilation of the function is postponed till S is known, ie. the compilation has a dependency on knowing S. 

Ideally in the future we would do a similar thing for GasMeter (it requires more thought because you don't want dynamic dispatch for each instruction) in which case the interpreter loop would be a regular function and compilation would happen completely at crate building time and not at monomorphizing time

Reply:
oh i see you're talking about compilation dependencies. yes impl is basically the same as generic static dispatch compared to dyn.